### PR TITLE
log changes

### PR DESCRIPTION
**Pull request by vaibhavsethia**
1. Changed revenueGoal to revenueValue ( Similar change needed in Example App)
2. Revenue Goal pushed was 5 because it was hardcoded, changed it to dynamic value
3. In eventDispatcher, Logged URL
4. Corrected Track comment for Goal id
5. Logged preSegmentation instead of preEvaluation
6. GetVariationName should only not work for FEATURE_ROLLOUT, changes done
7. UserStorage Log changed from ERROR to INFO, had to add a check for level 
8. Url Encode userID in impression call.

### CODE REVIEW

**Comment by vaibhavsethia** [comment]
Log level checks for FindTargettedVariation

File: pkg/api/getVariationName.go [hunk 1]
```diff
@@ -78,7 +78,7 @@ func (vwo *VWOInstance) GetVariationName(campaignKey, userID string, option inte
 		utils.LogMessage(vwo.Logger, constants.Error, getVariationName, message)
 		return ""
 	}
-	if !utils.CheckCampaignType(campaign, constants.CampaignTypeVisualAB) {
+	if utils.CheckCampaignType(campaign, constants.CampaignTypeFeatureRollout) {
```
Comment (Piyushhbhutoria) [inline pkg/api/getVariationName.go:81]: are you sure this new condition is right?

File: pkg/api/testdata/testdata.json [hunk 2]
```diff
@@ -22,7 +22,7 @@
       "isForcedVariationEnabled": false,
       "key": "CAMPAIGN_1",
       "variations": {},
-      "type": "FEATURE_TEST"
+      "type": "FEATURE_ROLLOUT"
```
Comment (Piyushhbhutoria) [inline pkg/api/testdata/testdata.json]: you are changing the test data, what about the test case?

File: pkg/constants/logMessages.go [hunk 3]
```diff
@@ -19,91 +19,92 @@ package constants
 // constants for logger
 const (
 	//Debug Messages
+	DebugMessageCustomLoggerUsed                = "Custom logger used"
 	DebugMessageDevelopmentMode                 = "Development mode is : %v "
+	DebugMessageGettingStoredVariation          = "[%v] Got stored variation for User ID: %v of Campaign: %v as Variation: %v, found in UserStorageService"
+	DebugMessageGotVariationForUser             = "[%v] User ID: %v of Campaign: %v campaignType: %v got variation: %v inside method: %v "
+	DebugMessageImpressionForPush               = "[%v] impression built for pushing - AccountID: %v, UserID: %v, SID: %v, URL: %v, U: %v "
+	DebugMessageImpressionForTrackCustomGoal   	= "[%v] impression built for track goal -  AccountID: %v, UserID: %v, SID: %v, URL: %v, ExperimentID: %v, Combination: %v, GoalID: %v "
+	DebugMessageImpressionForTrackRevenueGoal  	= "[%v] impression built for track goal -  AccountID: %v, UserID: %v, SID: %v, URL: %v, ExperimentID: %v, Combination: %v, GoalID: %v, RevenueValue: %v "
```
Comment (Piyushhbhutoria) [inline pkg/constants/logMessages.go]: use` %+v` instead of `%v` for a better output, do not pass so many values for no reason.

Comment (Piyushhbhutoria) [reply pkg/constants/logMessages.go]: same applies for most of the log messages here

File: pkg/constants/logMessages.go [hunk 4]
```diff
@@ -19,91 +19,92 @@ package constants
...
 	InfoMessageGotStoredVariation             = "[%v] Got stored variation: %v of CampaignKey: %v for User ID: %v from UserStorage"
-	InfoMessageSettingDataUserStorageService  = "[%v] Setting data into UserStorageService for User ID: %v successful"
+	InfoMessageGotVariationForUser            = "[%v] User ID: %v for CampaignKey: %v type: %v got variation_name: %v "
+	InfoMessageImpressionSuccess              = "[%v] Impression event - %v was successfully received by VWO having keys: %v "
 	InfoMessageInvalidVariationKey            = "[%v] Variation was not assigned to User ID: %v for Campaign: %v : %v "
-	InfoMessageVariationRangeAllocation       = "[%v] Variation: %v with weight: %v got range as: ( %v - %v )"
-	InfoMessageFeatureEnabledForUser          = "[%v] Campaign: %v for user ID: %v is enabled"
-	InfoMessageFeatureNotEnabledForUser       = "[%v] Campaign: %v for user ID: %v is not enabled"
+	InfoMessageMainKeysForPushAPI             = "[%v] Having main keys: AccountID: %v User ID: %v U: %v and tags: %v "
+	InfoMessageMainKeysForImpression          = "[%v] Having main keys: AccountID: %v User ID: %v campaignId: %v and VariationID: %v "
```
Comment (Piyushhbhutoria) [inline pkg/constants/logMessages.go]: use `%+v`

File: pkg/utils/impression.go [hunk 5]
```diff
@@ -101,7 +108,7 @@ func CreateImpressionTrackingUser(vwoInstance schema.VwoInstance, campaignID int
 	impression.ED = `{\"p\":\"` + constants.Platform + `\"}`
 	impression.URL = constants.HTTPSProtocol + constants.EndPointsBaseURL + constants.EndPointsTrackUser
 
-	message := fmt.Sprintf(constants.DebugMessageImpressionForTrackUser, vwoInstance.API, impression)
+	message := fmt.Sprintf(constants.DebugMessageImpressionForTrackUser, vwoInstance.API, impression.AccountID, impression.UID, impression.SID, impression.URL, impression.ExperimentID, impression.Combination, impression.ED)
```
Comment (Piyushhbhutoria) [inline pkg/utils/impression.go:118]: no need to send each attribute seperatly. use `%+v` in logs

File: pkg/utils/impression.go [hunk 6]
```diff
@@ -124,6 +131,6 @@ func getCommonProperties(vwoInstance schema.VwoInstance, userID string) schema.I
 		SID:       strconv.FormatInt(time.Now().Unix(), 10),
 		U:         generateFor(vwoInstance, userID, vwoInstance.SettingsFile.AccountID),
 		AccountID: vwoInstance.SettingsFile.AccountID,
-		UID:       userID,
+		UID:       url.QueryEscape(userID),
```
Comment (Piyushhbhutoria) [inline pkg/utils/impression.go]: is this working fine?

File: pkg/utils/impression.go [hunk 7]
```diff
@@ -72,12 +72,19 @@ func CreateImpressionTrackingGoal(vwoInstance schema.VwoInstance, variationID in
 
 	impression.URL = constants.HTTPSProtocol + constants.EndPointsBaseURL + constants.EndPointsTrackGoal
 	impression.GoalID = goalID
-	if revenueGoal > 0 {
-		impression.R = revenueGoal
+
+	if goalType==constants.GoalTypeRevenue && revenueValue >= 0 {
```
Comment (Piyushhbhutoria) [inline pkg/utils/impression.go]: what if this is false?

File: pkg/utils/impression.go [hunk 8]
```diff
@@ -46,21 +46,21 @@ func CreateImpressionForPush(vwoInstance schema.VwoInstance, tagKey, tagValue, u
 	parameters.Add(tagKey, tagValue)
 	impression.U = parameters.Encode()
 
-	message := fmt.Sprintf(constants.DebugMessageImpressionForPush, vwoInstance.API, impression)
+	message := fmt.Sprintf(constants.DebugMessageImpressionForPush, vwoInstance.API, impression.AccountID, impression.UID, impression.SID, impression.URL, impression.U)
```
Comment (Piyushhbhutoria) [inline pkg/utils/impression.go]: no need to send seperately.

File: pkg/event/eventDispatcher.go [hunk 9]
```diff
@@ -54,7 +54,7 @@ func Dispatch(vwoInstance schema.VwoInstance, impression schema.Impression) {
 			message := fmt.Sprintf(constants.ErrorMessageImpressionFailed, vwoInstance.API, err)
 			utils.LogMessage(vwoInstance.Logger, constants.Error, eventDispatcher, message)
 		} else {
-			message := fmt.Sprintf(constants.InfoMessageImpressionSuccess, vwoInstance.API, "Normal", impression)
+			message := fmt.Sprintf(constants.InfoMessageImpressionSuccess, vwoInstance.API, "Normal", URL)
```
Comment (Piyushhbhutoria) [inline pkg/event/eventDispatcher.go:61]: is this right?

File: pkg/core/variationDecider_test.go [hunk 10]
```diff
@@ -194,7 +194,7 @@ func TestGetVariationFromUserStorage(t *testing.T) {
 
 	campaign := vwoInstance.SettingsFile.Campaigns[0]
 	userID := "USER_3"
-	actual, err := GetVariationFromUserStorage(vwoInstance, userID, campaign)
+	actual, _, err := GetVariationFromUserStorage(vwoInstance, userID, campaign)
```
Comment (Piyushhbhutoria) [inline pkg/core/variationDecider_test.go]: do not ignore values in the test file. write an assert for the same.

File: pkg/core/variationDecider_test.go [hunk 11]
```diff
@@ -97,38 +97,38 @@ func TestFindTargetedVariation(t *testing.T) {
...
 	options := schema.Options{
 		VariationTargetingVariables: map[string]interface{}{"_vwo_user_id": "USER_1"},
-		RevenueGoal:     12,
+		RevenueValue:     12,
 	}
 
 	userID := "USER_1"
 	campaign := vwoInstance.SettingsFile.Campaigns[2]
-	actual, _ := FindTargetedVariation(vwoInstance, userID, campaign, options)
+	actual, _, _ := FindTargetedVariation(vwoInstance, userID, campaign, options)
```
Comment (Piyushhbhutoria) [inline pkg/core/variationDecider_test.go]: do not ignore values in the test file. write an assert for the same.

File: pkg/core/variationDecider.go [hunk 12]
```diff
@@ -68,18 +68,22 @@ func GetVariation(vwoInstance schema.VwoInstance, userID string, campaign schema
...
-	} else {
+		utils.LogMessage(vwoInstance.Logger, constants.Info, variationDecider, err.Error())
+	}else {
 		message := fmt.Sprintf(constants.InfoMessageGotVariationForUser, vwoInstance.API, userID, campaign.Key, campaign.Type, targettedVariation.Name)
 		utils.LogMessage(vwoInstance.Logger, constants.Info, variationDecider, message)
 		return targettedVariation, nil
 	}
 
-	variationName, err := GetVariationFromUserStorage(vwoInstance, userID, campaign)
+	variationName, level, err := GetVariationFromUserStorage(vwoInstance, userID, campaign)
```
Comment (Piyushhbhutoria) [inline pkg/core/variationDecider.go]: why do you need level? can't we do this in the function itself?

File: pkg/api/getVariationName.go (hunk 1 above)
Comment (vaibhavsethia) [reply pkg/api/getVariationName.go:81]: https://github.com/wingify/vwo-python-sdk/blob/master/vwo/api/get_variation_name.py
Line 78: 
If Campaign.Type is FEATURE_ROLLOUT then it should return else continue

**Review by Piyushhbhutoria — CHANGES_REQUESTED** [review]

File: pkg/api/testdata/testdata.json (hunk 2 above)
Comment (vaibhavsethia) [reply pkg/api/testdata/testdata.json]: Checked

File: pkg/utils/impression.go (hunk 5 above)
Comment (vaibhavsethia) [reply pkg/utils/impression.go:118]: Ok!

File: pkg/event/eventDispatcher.go (hunk 9 above)
Comment (vaibhavsethia) [reply pkg/event/eventDispatcher.go:61]: Earlier we were printing the whole impression, in other SDK they have just logged the endpoint (URL)

File: pkg/core/variationDecider.go (hunk 12 above)
Comment (vaibhavsethia) [reply pkg/core/variationDecider.go]: If we log the issue in the function itself, we wont be able to handle the cases in the calling function because there we are checking if the err := nil

Comment (vaibhavsethia) [reply pkg/core/variationDecider.go]: Or we can return some dummy error and just handle the case with logging in the function not in the calling function

File: pkg/core/variationDecider_test.go (hunk 11 above)
Comment (vaibhavsethia) [reply pkg/core/variationDecider_test.go]: Ok

File: pkg/core/variationDecider_test.go (hunk 10 above)
Comment (vaibhavsethia) [reply pkg/core/variationDecider_test.go]: Ok

File: pkg/utils/impression.go (hunk 8 above)
Comment (vaibhavsethia) [reply pkg/utils/impression.go]: Will %+v log all values along with the names , Like Key : value pair ?

File: pkg/utils/impression.go (hunk 7 above)
Comment (vaibhavsethia) [reply pkg/utils/impression.go]: If GoalType is not REVENUE_TRACKING we dont need to address impression.R

File: pkg/utils/impression.go (hunk 6 above)
Comment (vaibhavsethia) [reply pkg/utils/impression.go]: Yes

File: pkg/core/variationDecider_test.go [hunk 13]
```diff
@@ -97,40 +98,49 @@ func TestFindTargetedVariation(t *testing.T) {
...
-	actual, _ = FindTargetedVariation(vwoInstance, userID, campaign, options)
+	actual, level, err = FindTargetedVariation(vwoInstance, userID, campaign, options)
 	assertOutput.Empty(actual, "Expected no variation")
+	assertOutput.Equal(level, constants.Info, "Mismatching Log Level")
+	assertOutput.NotNil(err, "Error Expected")
 
 	userID = "USER_1"
 	campaign = vwoInstance.SettingsFile.Campaigns[4]
-	actual, _ = FindTargetedVariation(vwoInstance, userID, campaign, options)
+	actual, level, err = FindTargetedVariation(vwoInstance, userID, campaign, options)
```
Comment (Piyushhbhutoria) [inline pkg/core/variationDecider_test.go]: isn't level removed yet?

**Review by Piyushhbhutoria — CHANGES_REQUESTED** [review]

**Review by Piyushhbhutoria — APPROVED** [review]

File: pkg/api/vwo.go [hunk 14]
```diff
@@ -88,6 +88,6 @@ func WithLogger(logger interface{}) VWOOption {
 // WithDevelopmentMode sets development mode true
 func WithDevelopmentMode() VWOOption {
 	return func(vwo *VWOInstance) {
-		vwo.IsDevelopmentMode = true
+		vwo.IsDevelopmentMode = false
```
Comment (Piyushhbhutoria) [inline pkg/api/vwo.go]: this is wrong, undo this.

**Review by Piyushhbhutoria — CHANGES_REQUESTED** [review]

File: go.sum [hunk 15]
```diff
@@ -2,6 +2,7 @@ github.com/davecgh/go-spew v1.1.0 h1:ZDRjVQ15GmhC3fiQ8ni8+OwkZQO4DARzQgrnXU1Liz8
 github.com/davecgh/go-spew v1.1.0/go.mod h1:J7Y8YcW2NihsgmVo/mv3lAwl/skON4iLHjSsI+c5H38=
 github.com/google/uuid v1.1.1 h1:Gkbcsh/GbpXz7lPftLA3P6TYMwjCLYm83jiFQZF/3gY=
 github.com/google/uuid v1.1.1/go.mod h1:TIyPZe4MgqvfeYDBFedMoGGpEw/LqOeaOT+nhxU+yHo=
+github.com/pmezard/go-difflib v1.0.0 h1:4DBwDE0NGyQoBHbLQYPwSUPoCMWR5BEzIk/f1lZbAQM=
```
Comment (Piyushhbhutoria) [inline go.sum:12]: why this change?
